### PR TITLE
Add new 'MinimalTest' suite to compilertest.

### PR DESCRIPTION
**Pull request by janvrany**
This suite contains a really minimal tests such as "return 42"
or "return arg + 10". These tests are useful when porting compiler
to a new architecture, such as RISC-V.

### CODE REVIEW

**Comment by janvrany** [comment]
Ah, I did not know `compilertest` is being deprecated. It can be moved to TRIL tests, the reason we did it this way is that `compilertest` is easier to debug (or it seems when we started with RISC-V port). 

I'm fine to move it to TRIL tests or drop it altogether, whatever you see fit.

**Comment by Leonardo2718** [comment]
I think the tests you have are already covered by some combination of JitBuilder and Tril tests. There might be some cases missing in the Tril tests, though. I'll let you decide if you want to just drop these or add equivalent Tril tests for the uncovered cases. :slightly_smiling_face:

**Comment by shingarov** [comment]
> tests you have are already covered by some combination of JitBuilder and Tril tests

**Comment by janvrany** [comment]
@shingarov 
> would certainly be way too complex and would lead to guaranteed failure. I needed something much more simple.
Yes. On the other hand, my undestanding is that back then part of the problem (most, maybe) was to get TRIL compiled as TRIL requires jitbuilder. Back then you used autotools and `make` and compiling jitbuilder using autotools and make was problematic. Do I recall it correctly? 

Today, however, CMake build is "the build" so compiling jitbuilder and tril is not an issue. In fact, it compiled out-of-the-box when I started the last iteration. So maybe, this is no longer an issue and having them moved over to TRIL is fine.

I do agree that the point of `MinimalTest` is not coverage, but super-simple set of tests to start with when porting to a new ISA.

**Comment by janvrany** [comment]
@Leonardo2718 : given that TRIL has no support for expressing recursive call, factorial / fibonacci tests can't be easily ported. These proved to be *very* useful when working on the port, so I'd like *not* to loose them. What do you suggest?

**Comment by janvrany** [comment]
@Leonardo2718 
>  I don't think Tril is going to get proper support for calls anytime soon, so putting those tests in the compilertriltest/ directory but leaving them as is should be OK.

I'm sorry, I'm little confused now. Are you suggesting to put them into `compilertriltest` but use JitBuilder API directly instead of TRIL?

**Comment by Leonardo2718** [comment]
@janvrany 

> I'm sorry, I'm little confused now. Are you suggesting to put them into compilertriltest but use JitBuilder API directly instead of TRIL?

Yes, my appologies for the lack of clarity. `TR::MethodBuilder` and the other builder classes can be used without the JitBuilder library. So, it is possible to use those APIs in the test compiler and Tril.

**Comment by shingarov** [comment]
Now I am even more confused, by the phrase "use JitBuilder".  I wrote these `ReturnArgI32`, `ArgConst` etc tests exactly because I could *not* use JitBuilder.  Actually, in the first cut of the RISC-V port, I could not even use GTest, so there were straight `main()` methods driving the examples.  If we bring the mention of "JitBuilder" into the picture, this kind of defeats the whole purpose of these tests.

Probably what is actually meant here, is the `TR::MethodBuilder` API, *not* JitBuilder?

**Comment by Leonardo2718** [comment]
@shingarov 

> Probably what is actually meant here, is the `TR::MethodBuilder` API, not JitBuilder?

Yes, that is correct. Usually, when talking about `MethodBuilder`, people mean the JitBuilder API. As I mentioned above, the API _can_ be used without the JitBuilder library, but its primarly intended for use with JitBuilder.

**Comment by janvrany** [comment]
@Leonardo2718 , @shingarov 

Okay, let me try again and then reiterate.

**Comment by janvrany** [comment]
Polite ping.

**Comment by fjeremic** [comment]
@Leonardo2718 do you want to sign up to be the committer for this one?

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+#include "compile/CompilationTypes.hpp"
+#include "il/DataTypes.hpp"
+#include "ilgen/IlGeneratorMethodDetails_inlines.hpp"
+#include "ilgen/TypeDictionary.hpp"
+#include "ilgen/MethodBuilder.hpp"
+
+class MinimalTest : public TRTest::JitTest {
+public:
+	MinimalTest();
+};
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]: Sorry to come in late with a nitpick, but this class doesn't follow the jit coding style.

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+	int32_t rc = 0;
+	F entry = (F)compileMethodFromDetails(NULL, details, warm, rc);
+	return entry;
+}
+
+#define COMPILE(test, name) { \
+   TR::TypeDictionary types; \
+   name ## Method builder(&types); \
+   entry = compile<decltype(entry)>(&builder); \
+   }
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]: I'm personally biased against using macros, and I would prefer if you just wrote this code out by hand, or used helper functions to reduce the boilerplate.

Comment (janvrany) [reply fvtest/compilertriltest/MinimalTest.cpp]: @rwy0717: I rewrote it using templates and it looks much nicer to me. Thanks for the suggestion!

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+   DefineParameter("leftV", Int32);
+   DefineParameter("rightV", Int32);
+   DefineReturnType(Int32);
+   }
+
+bool
+MaxIfThenMethod::buildIL()
+   {
+   TR::IlBuilder *thenPath = NULL;
+   TR::IlValue *condition;
```
Comment (Leonardo2718) [inline fvtest/compilertriltest/MinimalTest.cpp]: This variable is unused.

**Comment by Leonardo2718** [comment]
@genie-omr build all

**Comment by janvrany** [comment]
@Leonardo2718 "Looks good", but it is not :-( The template magic does not compile, perhaps I'm using too new GCC. Will check. Sorry about that.

**Comment by rwy0717** [comment]
Error on AIX:
```
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 57.26: 1540-1118 (S) The declaration of "resolvedMethod" uses the undefined class "TR::ResolvedMethod" when the class must be complete.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 73.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int ()>" cannot be initialized because it does not have a default constructor.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 98.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int)>" cannot be initialized because it does not have a default constructor.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 127.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int, int)>" cannot be initialized because it does not have a default constructor.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 165.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int)>" cannot be initialized because it does not have a default constructor.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 193.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int, int)>" cannot be initialized because it does not have a default constructor.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 222.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int)>" cannot be initialized because it does not have a default constructor.
[2019-11-01T20:31:37.917Z] "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 267.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int)>" cannot be initialized because it does not have a default constructor.
```
And same error on x86-64 (this one is a little clearer):
```
[2019-11-01T20:27:47.477Z] /home/jenkins/workspace/PullRequest-linux_x86-64/fvtest/compilertriltest/MinimalTest.cpp: In instantiation of ‘S* StaticSignatureMethodBuilder<S>::Compile() [with S = int(); StaticSignatureMethodBuilder<S>::FunctionPtr = int (*)()]’:
[2019-11-01T20:27:47.477Z] /home/jenkins/workspace/PullRequest-linux_x86-64/fvtest/compilertriltest/MinimalTest.cpp:313:21:   required from ‘typename M::FunctionPtr compile() [with M = MeaningOfLifeMethod; typename M::FunctionPtr = int (*)()]’
[2019-11-01T20:27:47.477Z] /home/jenkins/workspace/PullRequest-linux_x86-64/fvtest/compilertriltest/MinimalTest.cpp:318:46:   required from here
[2019-11-01T20:27:47.477Z] /home/jenkins/workspace/PullRequest-linux_x86-64/fvtest/compilertriltest/MinimalTest.cpp:57:45: error: ‘TR::ResolvedMethod resolvedMethod’ has incomplete type
[2019-11-01T20:27:47.477Z]        TR::ResolvedMethod resolvedMethod(this);
```

Possibly, we're missing a header for the ResolvedMethod class?

**Comment by janvrany** [comment]
I rebased my PR on current master and got the same error - something has changed meanwhile which explains why I did not see that error when testing locally. Will fix soon.

**Comment by janvrany** [comment]
@genie-omr build all

**Comment by rwy0717** [comment]
@genie-omr build all

**Comment by janvrany** [comment]
OK, so we have two different problems: 

 * On AIX / zOS it fails to compile with `"/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 98.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int)>" cannot be initialized because it does not have a default constructor.` Not exactly sure what this does mean. I'd appreciate any help from someone more knowledgeable and with an access xOS / AIX with XLC compiler. 

* On Linux on PPC, `MinimalTest` segfaults. I believe this is because the `PPCSystemLinkage` is broken (as @shingarov found out earlier). Cannot check right now as I don't have Linux on PPC system set up. @Leonardo2718 @rwy0717, what do you suggest to do about it?

**Comment by Leonardo2718** [comment]
> * On AIX / zOS it fails to compile with "/home/u0020236/workspace/PullRequest-aix_ppc-64/fvtest/compilertriltest/MinimalTest.cpp", line 98.4: 1540-0239 (S) The base class "StaticSignatureMethodBuilder<int (int)>" cannot be initialized because it does not have a default constructor. Not exactly sure what this does mean. I'd appreciate any help from someone more knowledgeable and with an access xOS / AIX with XLC compiler.

I believe this is an xlc bug. Can you try using function pointers instead?

> * On Linux on PPC, MinimalTest segfaults. I believe this is because the PPCSystemLinkage is broken (as @shingarov found out earlier). Cannot check right now as I don't have Linux on PPC system set up. @Leonardo2718 @rwy0717, what do you suggest to do about it?

You should be able to skip the failing test using the `SKIP_IF` macro, e.g. https://github.com/eclipse/omr/blob/2c18aa559c92701900485bfdb61438b8b4e095b4/fvtest/compilertriltest/ArithmeticTest.cpp#L145-L146

Is there an issue already open about the PPCSystemLinkage problem?

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+
+class MeaningOfLifeMethod : public StaticSignatureMethodBuilder<int32_t()>
+   {
+   public:
+   MeaningOfLifeMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+MeaningOfLifeMethod::MeaningOfLifeMethod(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t()>(types)
```

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+class ReturnArgI32Method : public StaticSignatureMethodBuilder<int32_t(int32_t)>
+   {
+   public:
+   ReturnArgI32Method(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+ReturnArgI32Method::ReturnArgI32Method(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t(int32_t)>(types)
```

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+
+class MaxIfThenMethod : public StaticSignatureMethodBuilder<int32_t(int32_t, int32_t)>
+   {
+   public:
+   MaxIfThenMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+MaxIfThenMethod::MaxIfThenMethod(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t(int32_t, int32_t)>(types)
```

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+
+class AddArgConstMethod : public StaticSignatureMethodBuilder<int32_t(int32_t)>
+   {
+   public:
+   AddArgConstMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+AddArgConstMethod::AddArgConstMethod(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t(int32_t)>(types)
```

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+class SubArgArgMethod : public StaticSignatureMethodBuilder<int32_t(int32_t, int32_t)>
+   {
+   public:
+   SubArgArgMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+SubArgArgMethod::SubArgArgMethod(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t(int32_t, int32_t)>(types)
```

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+
+class FactorialMethod : public StaticSignatureMethodBuilder<int32_t(int32_t)>
+   {
+   public:
+   FactorialMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+FactorialMethod::FactorialMethod(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t(int32_t)>(types)
```

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+class RecursiveFibonnaciMethod : public StaticSignatureMethodBuilder<int32_t(int32_t)>
+   {
+   public:
+	RecursiveFibonnaciMethod(TR::TypeDictionary *types);
+   virtual bool buildIL();
+   };
+
+
+RecursiveFibonnaciMethod::RecursiveFibonnaciMethod(TR::TypeDictionary *types)
+   : StaticSignatureMethodBuilder(types)
```
Comment (rwy0717) [inline fvtest/compilertriltest/MinimalTest.cpp]:
```suggestion
   : StaticSignatureMethodBuilder<int32_t(int32_t)>(types)
```

**Comment by janvrany** [comment]
Sorry for late reply, I was busy with conference stuff. 
@rwy0717 : Thanks, I did as you suggested, let's see if it helps.
@Leonardo2718: Okay, I added `SKIP_IF` to skip tests on power. As for the issue, I think this is covered #1645, I'm not aware of specific issue for this. I'll talk to @shingarov (as he knows details) and we'll open an issue.

**Comment by rwy0717** [comment]
@genie-omr build all

**Comment by Leonardo2718** [comment]
The added tests are failing on AIX. Curious.

**Comment by janvrany** [comment]
@Leonardo2718 : Yeah. I thought this is because of the linkage but Linux / PPC seem to work. Maybe it still is because PPC code is sprinkled with `isAIXstyleLinkage` or alike. 
Unfortunately I do not have PPC + AIX system so all I can do is either gaze to crystal ball or `SKIP_IF(...AIX...)` 

So unless anyone objects, I'll skip the test on AIX and open an issue.

**Comment by rwy0717** [comment]
@genie-omr build all

**Comment by janvrany** [comment]
Okay, so `MinimalTest` is disabled on POWER until further investigation. All jobs are green. Is there anything else I shall improve?

File: fvtest/compilertriltest/MinimalTest.cpp [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+   TR::TypeDictionary types;
+   M m(&types);
+   return m.Compile();
+   }
+
+TEST_F(MinimalTest, MeaningOfLife)
+   {
+   std::string arch = omrsysinfo_get_CPU_architecture();
+   SKIP_IF((OMRPORT_ARCH_PPC == arch || OMRPORT_ARCH_PPC64 == arch || OMRPORT_ARCH_PPC64LE == arch), MissingImplementation)
+           << "Test is skipped on non-x86-64 platforms because calls are not currently supported on other POWER (see issue #1645)";
```
Comment (Leonardo2718) [inline fvtest/compilertriltest/MinimalTest.cpp]: I don't think the message is accurate since the test does run on some non-x86-64 platforms (and the issue referenced is not related).

**Comment by fjeremic** [comment]
@genie-omr build all

**Comment by Leonardo2718** [comment]
Changes are approved and tests have passed. Merging.
